File: 14-TypeScript/Grey Multi-Tenant/packages/grey-matlab/+grey/+error/ErrorCodes.m
classdef ErrorCodes
    %ERRORCODES Error code constants for Grey SDK
    %   Defines standard error codes used throughout the SDK.
    
    properties (Constant)
        UNAUTHORIZED = 'unauthorized'
        FORBIDDEN = 'forbidden'
        NOT_FOUND = 'not_found'
        VALIDATION_ERROR = 'validation_error'
        NETWORK_ERROR = 'network_error'
        TIMEOUT = 'timeout'
        SERVER_ERROR = 'server_error'
        UNKNOWN = 'unknown'
    end
    
    methods (Static)
        function code = fromHttpStatus(status)
            %FROMHTTPSTATUS Convert HTTP status code to error code
            %   code = ErrorCodes.fromHttpStatus(status)
            
            if status == 401
                code = grey.error.ErrorCodes.UNAUTHORIZED;
            elseif status == 403
                code = grey.error.ErrorCodes.FORBIDDEN;
            elseif status == 404
                code = grey.error.ErrorCodes.NOT_FOUND;
            elseif status == 400 || status == 422
                code = grey.error.ErrorCodes.VALIDATION_ERROR;
            elseif status == 408 || status == 504
                code = grey.error.ErrorCodes.TIMEOUT;
            elseif status >= 500 && status < 600
                code = grey.error.ErrorCodes.SERVER_ERROR;
            else
                code = grey.error.ErrorCodes.UNKNOWN;
            end
        end
        
        function valid = isValid(code)
            %ISVALID Check if error code is valid
            %   valid = ErrorCodes.isValid(code)
            
            validCodes = {
                grey.error.ErrorCodes.UNAUTHORIZED, ...
                grey.error.ErrorCodes.FORBIDDEN, ...
                grey.error.ErrorCodes.NOT_FOUND, ...
                grey.error.ErrorCodes.VALIDATION_ERROR, ...
                grey.error.ErrorCodes.NETWORK_ERROR, ...
                grey.error.ErrorCodes.TIMEOUT, ...
                grey.error.ErrorCodes.SERVER_ERROR, ...
                grey.error.ErrorCodes.UNKNOWN
            };
            
            valid = any(strcmp(code, validCodes));
        end
        
        function normalized = normalize(code)
            %NORMALIZE Normalize error code to valid value
            %   normalized = ErrorCodes.normalize(code)
            
            if grey.error.ErrorCodes.isValid(code)
                normalized = code;
            else
                normalized = grey.error.ErrorCodes.UNKNOWN;
            end
        end
    end
end

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-julia/src/ErrorCodes.jl
"""
Error codes for the Grey SDK.
"""
module ErrorCodes

export ErrorCode, 
       UNAUTHORIZED, FORBIDDEN, NOT_FOUND, VALIDATION_ERROR,
       NETWORK_ERROR, TIMEOUT, SERVER_ERROR, UNKNOWN,
       from_http_status, is_valid, normalize

"""Error code type"""
const ErrorCode = Symbol

# Error code constants
const UNAUTHORIZED = :unauthorized
const FORBIDDEN = :forbidden
const NOT_FOUND = :not_found
const VALIDATION_ERROR = :validation_error
const NETWORK_ERROR = :network_error
const TIMEOUT = :timeout
const SERVER_ERROR = :server_error
const UNKNOWN = :unknown

# All valid error codes
const VALID_CODES = Set([
    UNAUTHORIZED, FORBIDDEN, NOT_FOUND, VALIDATION_ERROR,
    NETWORK_ERROR, TIMEOUT, SERVER_ERROR, UNKNOWN
])

"""
    from_http_status(status::Int) -> ErrorCode

Convert an HTTP status code to an error code.
"""
function from_http_status(status::Int)::ErrorCode
    if status == 401
        return UNAUTHORIZED
    elseif status == 403
        return FORBIDDEN
    elseif status == 404
        return NOT_FOUND
    elseif status in (400, 422)
        return VALIDATION_ERROR
    elseif status in (408, 504)
        return TIMEOUT
    elseif 500 <= status < 600
        return SERVER_ERROR
    else
        return UNKNOWN
    end
end

"""
    is_valid(code::ErrorCode) -> Bool

Check if an error code is valid.
"""
function is_valid(code::ErrorCode)::Bool
    code in VALID_CODES
end

"""
    normalize(code::ErrorCode) -> ErrorCode

Normalize an error code to a valid value.
"""
function normalize(code::ErrorCode)::ErrorCode
    is_valid(code) ? code : UNKNOWN
end

end # module
